homeassistant/components/prometheus/__init__.py

```diff
@@ -274,15 +286,25 @@ def handle_entity_registry_updated(
             self._remove_labelsets(metrics_entity_id)
 
     def _remove_labelsets(
-        self, entity_id: str, friendly_name: str | None = None
+        self,
+        entity_id: str,
+        friendly_name: str | None = None,
+        ignored_metric_names: set[str] | None = None,
     ) -> None:
         """Remove labelsets matching the given entity id from all metrics."""
+        if ignored_metric_names is None:
+            ignored_metric_names = set()
         for metric in list(self._metrics.values()):
             for sample in cast(list[prometheus_client.Metric], metric.collect())[
                 0
             ].samples:
-                if sample.labels["entity"] == entity_id and (
-                    not friendly_name or sample.labels["friendly_name"] == friendly_name
+                if (
+                    sample.name not in ignored_metric_names
```

Comment:
Could we skip based on the key in `self._metrics`? It's supposed to be the metric name anyway.

Reply:
Yes, this makes things much better. Thanks!